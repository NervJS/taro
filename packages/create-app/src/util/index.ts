import * as path from 'node:path'

export function getRootPath (): string {
  return path.resolve(__dirname, '../../')
}

export function getPkgVersion (): string {
  return require(path.join(getRootPath(), 'package.json')).version
}

export function printPkgVersion () {
  console.log(`ðŸ‘½ Taro v${getPkgVersion()}`)
  console.log()
}
