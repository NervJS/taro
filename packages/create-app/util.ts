import { version } from './package.json'

export function printPkgVersion () {
  console.log(`ðŸ‘½ Taro v${version}`)
  console.log()
}
