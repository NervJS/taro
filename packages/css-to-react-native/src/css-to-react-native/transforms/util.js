import { tokens } from '../tokenTypes'

const { LENGTH, UNSUPPORTED_LENGTH_UNIT, PERCENT, COLOR, SPACE, NONE } = tokens

export const directionFactory = ({
  types = [LENGTH, UNSUPPORTED_LENGTH_UNIT, PERCENT],
  directions = ['Top', 'Right', 'Bottom', 'Left'],
  prefix = '',
  suffix = ''
}) => tokenStream => {
  const values = []

  // borderWidth doesn't currently allow a percent value, but may do in the future
  values.push(tokenStream.expect(...types))

  while (values.length < 4 && tokenStream.hasTokens()) {
    tokenStream.expect(SPACE)
    values.push(tokenStream.expect(...types))
  }

  tokenStream.expectEmpty()

  const [top, right = top, bottom = top, left = right] = values

  const keyFor = n => `${prefix}${directions[n]}${suffix}`

  const output = {
    [keyFor(0)]: top,
    [keyFor(1)]: right,
    [keyFor(2)]: bottom,
    [keyFor(3)]: left
  }

  return { $merge: output }
}

export const anyOrderFactory = (properties, delim = SPACE) => tokenStream => {
  const propertyNames = Object.keys(properties)
  const values = propertyNames.reduce((accum, propertyName) => {
    accum[propertyName] === undefined; // eslint-disable-line
    return accum
  }, {})

  let numParsed = 0
  while (numParsed < propertyNames.length && tokenStream.hasTokens()) {
    if (numParsed) tokenStream.expect(delim)

    const matchedPropertyName = propertyNames.find(
      propertyName =>
        values[propertyName] === undefined &&
        properties[propertyName].tokens.some(token =>
          tokenStream.matches(token)
        )
    )

    if (!matchedPropertyName) {
      tokenStream.throw()
    } else {
      values[matchedPropertyName] = tokenStream.lastValue
    }

    numParsed += 1
  }

  tokenStream.expectEmpty()

  propertyNames.forEach(propertyName => {
    if (values[propertyName] === undefined) { values[propertyName] = properties[propertyName].default }
  })

  return { $merge: values }
}

export const shadowOffsetFactory = () => tokenStream => {
  const width = tokenStream.expect(LENGTH)
  const height = tokenStream.matches(SPACE)
    ? tokenStream.expect(LENGTH)
    : width
  tokenStream.expectEmpty()
  return { width, height }
}

export const parseShadow = tokenStream => {
  let offsetX
  let offsetY
  let radius
  let color

  if (tokenStream.matches(NONE)) {
    tokenStream.expectEmpty()
    return {
      offset: { width: 0, height: 0 },
      radius: 0,
      color: 'black'
    }
  }

  let didParseFirst = false
  while (tokenStream.hasTokens()) {
    if (didParseFirst) tokenStream.expect(SPACE)

    if (
      offsetX === undefined &&
      tokenStream.matches(LENGTH, UNSUPPORTED_LENGTH_UNIT)
    ) {
      offsetX = tokenStream.lastValue
      tokenStream.expect(SPACE)
      offsetY = tokenStream.expect(LENGTH, UNSUPPORTED_LENGTH_UNIT)

      tokenStream.saveRewindPoint()
      if (
        tokenStream.matches(SPACE) &&
        tokenStream.matches(LENGTH, UNSUPPORTED_LENGTH_UNIT)
      ) {
        radius = tokenStream.lastValue
      } else {
        tokenStream.rewind()
      }
    } else if (color === undefined && tokenStream.matches(COLOR)) {
      color = tokenStream.lastValue
    } else {
      tokenStream.throw()
    }

    didParseFirst = true
  }

  if (offsetX === undefined) tokenStream.throw()

  return {
    offset: { width: offsetX, height: offsetY },
    radius: radius !== undefined ? radius : 0,
    color: color !== undefined ? color : 'black'
  }
}
