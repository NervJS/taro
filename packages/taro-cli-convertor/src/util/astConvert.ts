import generate from '@babel/generator'
import { NodePath } from '@babel/traverse'
import * as t from '@babel/types'

// 最低限度的转义： https://github.com/mathiasbynens/jsesc#minimal
export function generateMinimalEscapeCode (ast: t.File) {
  return generate(ast as any, {
    jsescOption: {
      minimal: true,
    },
  }).code
}

// 判断是否已经引入 @tarojs/taro
export function hasTaroImport (bodyNode: NodePath<t.Node>[]) {
  return bodyNode.some((astPath) => {
    // import Taro from '@tarojs/taro'
    if (astPath.isImportDeclaration()) {
      return astPath.node.source.value === '@tarojs/taro'
    } else if (astPath.isVariableDeclaration() && astPath.node.declarations.length === 1) {
      // const Taro = require('@tarojs/taro')
      const declaration: t.VariableDeclarator = astPath.node.declarations[0]
      return (
        declaration.init != null &&
        declaration.init.type === 'CallExpression' &&
        declaration.init.callee.type === 'Identifier' &&
        declaration.init.callee.name === 'require' &&
        declaration.init.arguments.length === 1 &&
        declaration.init.arguments[0].type === 'StringLiteral' &&
        declaration.init.arguments[0].value === '@tarojs/taro'
      )
    }
    return false
  })
}

// 根据关键字exports判断是否为commonjs模块
export function isCommonjsModule (bodyNode: NodePath<t.Node>[]) {
  return bodyNode.some((p) => {
    if (t.isExpressionStatement(p) && t.isAssignmentExpression(p.expression)) {
      const expression = p.expression
      // 1、module.exports.num = num 2、module.exports = {}
      const isModuleExports =
        (expression.left.type === 'MemberExpression' &&
          expression.left.object.type === 'MemberExpression' &&
          expression.left.object.property.type === 'Identifier' &&
          expression.left.object.property.name === 'exports') ||
        (expression.left.type === 'MemberExpression' &&
          expression.left.property.type === 'Identifier' &&
          expression.left.property.name === 'exports')

      // exports.num = num
      const isExports =
        expression.left.type === 'MemberExpression' &&
        expression.left.object.type === 'Identifier' &&
        expression.left.object.name === 'exports'
      return isModuleExports || isExports
    }
    return false
  })
}

/**
 * 判断节点是否为commonjs的导入
 *
 * @param {NodePath<t.Node>} bodyNode
 * @returns {boolean}
 */
export function isCommonjsImport (bodyNode: NodePath<t.Node>) {
  if (bodyNode.isVariableDeclaration() && bodyNode.node.declarations.length === 1) {
    const declaration: t.VariableDeclarator = bodyNode.node.declarations[0]
    return (
      declaration.init != null &&
      declaration.init.type === 'CallExpression' &&
      declaration.init.callee.type === 'Identifier' &&
      declaration.init.callee.name === 'require'
    )
  }
  return false
}
