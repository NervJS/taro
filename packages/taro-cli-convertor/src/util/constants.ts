export const enum Constants {
  PLUGIN = 'plugin',

  PLUGIN_JSON = 'plugin.json',
}
