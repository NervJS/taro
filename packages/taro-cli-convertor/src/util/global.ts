import { ErrCodeMsg } from './index'

// taro-cli-convertor 模块的全局变量
interface Globals {
  logFilePath: string   // 日志文件路径
  errCodeMsgs: ErrCodeMsg[]   // 错误代码信息 (转换报告涉及的字段)
  currentParseFile: string    // 当前解析的文件 (转换报告涉及的字段)
  logCount: number 
  logFileContent: string
}

export const globals: Globals = {
  logFilePath: '',
  errCodeMsgs: [],
  currentParseFile: '',
  logCount: 0,
  logFileContent: ''
}
