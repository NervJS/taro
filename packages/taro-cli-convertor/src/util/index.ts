import {
  chalk,
  CSS_IMPORT_REG,
  fs,
  printLog,
  processTypeEnum,
  promoteRelativePath,
  REG_SCRIPT,
  REG_TYPESCRIPT,
  resolveScriptPath,
  SCRIPT_EXT,
} from '@tarojs/helper'
import * as path from 'path'

import type * as t from '@babel/types'

const NODE_MODULES = 'node_modules'

interface IReportMsg {
  filePath: string // 报告信息文件路径
  message: string // 报告信息
  type?: string // 报告信息类型
  childReportMsg?: IReportMsg[]
}

export function getRootPath (): string {
  return path.resolve(__dirname, '../../')
}

export function getPkgVersion (): string {
  return require(path.join(getRootPath(), 'package.json')).version
}

// 文件存在或添加后缀.js、.jsx、.ts、.tsx存在则返回文件路径，否则返回null
function revertScriptPath (absolutePath: string, SCRIPT_EXT: string[]) {
  for (const item of SCRIPT_EXT) {
    if (fs.existsSync(absolutePath)) {
      return absolutePath
    } else if (fs.existsSync(`${absolutePath}${item}`)) {
      return `${absolutePath}${item}`
    }
  }
  return null
}

function getRelativePath (_rootPath: string, sourceFilePath: string, oriPath: string) {
  // 处理以/开头的绝对路径，比如 /a/b
  if (path.isAbsolute(oriPath)) {
    if (oriPath.indexOf('/') !== 0) {
      return ''
    }
    const absolutePath = revertScriptPath(path.resolve(sourceFilePath, '..' + oriPath), SCRIPT_EXT)
    if (absolutePath == null) {
      return ''
    }

    let relativePath = path.relative(path.dirname(sourceFilePath), absolutePath)
    relativePath = promoteRelativePath(relativePath)
    if (relativePath.indexOf('.') !== 0) {
      return `./${relativePath}`
    }
    return relativePath
  }
  // 处理非正常路径，比如 a/b
  if (oriPath.indexOf('.') !== 0) {
    const absolutePath = revertScriptPath(path.resolve(sourceFilePath, '..', oriPath), SCRIPT_EXT)

    // 可能为三方库
    if (absolutePath == null) {
      return oriPath
    }

    let relativePath = path.relative(path.dirname(sourceFilePath), absolutePath)
    relativePath = promoteRelativePath(relativePath)
    if (relativePath.indexOf('.') !== 0) {
      return `./${relativePath}`
    }
    return relativePath
  }
  return oriPath
}

export function copyFileToTaro (from: string, to: string, options?: fs.CopyOptionsSync) {
  const filename = path.basename(from)
  if (fs.statSync(from).isFile() && !path.extname(to)) {
    fs.ensureDir(to)
    return fs.copySync(from, path.join(to, filename), options)
  }
  fs.ensureDir(path.dirname(to))
  return fs.copySync(from, to, options)
}

// 递归复制文件夹
function copyFolderToTaro (source: string, destination: string) {
  // 创建目标文件夹
  if (!fs.existsSync(destination)) {
    fs.mkdirSync(destination, { recursive: true })
  }

  // 获取源文件夹中的所有项
  const items = fs.readdirSync(source)
  for (const item of items) {
    const sourcePath = path.join(source, item)
    const destinationPath = path.join(destination, item)

    const stat = fs.lstatSync(sourcePath)
    if (stat.isSymbolicLink()) {
      // 如果是符号链接，获取链接的目标路径并进行复制
      const target = fs.readlinkSync(sourcePath)
      copyFolderToTaro(target, destinationPath)
    } else if (stat.isDirectory()) {
      // 如果是子文件夹，则递归复制
      copyFolderToTaro(sourcePath, destinationPath)
    } else {
      // 如果是普通文件，则直接复制
      fs.copyFileSync(sourcePath, destinationPath)
    }
  }
}

export function analyzeImportUrl (
  rootPath: string,
  sourceFilePath: string,
  scriptFiles: Set<string>,
  source: t.StringLiteral,
  value: string,
  isTsProject?: boolean
) {
  const valueExtname = path.extname(value)
  const rpath = getRelativePath(rootPath, sourceFilePath, value)
  if (!rpath) {
    printLog(processTypeEnum.ERROR, '引用文件', `文件 ${sourceFilePath} 中引用 ${value} 不存在！`)
    return
  }
  if (rpath !== value) {
    value = rpath
    source.value = rpath
  }
  if (value.indexOf('.') === 0) {
    if (REG_SCRIPT.test(valueExtname) || REG_TYPESCRIPT.test(valueExtname)) {
      const vpath = path.resolve(sourceFilePath, '..', value)
      let fPath = value
      if (fs.existsSync(vpath)) {
        fPath = vpath
      } else {
        printLog(processTypeEnum.ERROR, '引用文件', `文件 ${sourceFilePath} 中引用 ${value} 不存在！`)
      }
      scriptFiles.add(fPath)
    } else {
      let vpath = resolveScriptPath(path.resolve(sourceFilePath, '..', value))
      if (vpath) {
        if (!fs.existsSync(vpath)) {
          printLog(processTypeEnum.ERROR, '引用文件', `文件 ${sourceFilePath} 中引用 ${value} 不存在！`)
        } else {
          if (fs.lstatSync(vpath).isDirectory()) {
            if (fs.existsSync(path.join(vpath, 'index.js'))) {
              vpath = path.join(vpath, 'index.js')
            } else {
              printLog(processTypeEnum.ERROR, '引用目录', `文件 ${sourceFilePath} 中引用了目录 ${value}！`)
              return
            }
          }
          let relativePath = path.relative(sourceFilePath, vpath)
          const relativePathExtname = path.extname(relativePath)
          scriptFiles.add(vpath)
          relativePath = promoteRelativePath(relativePath)
          if (/\.wxs/.test(relativePathExtname)) {
            relativePath += '.js'
          } else if (!isTsProject) {
            relativePath = relativePath.replace(relativePathExtname, '.js')
          }
          source.value = relativePath
        }
      }
    }
  } else {
    if (value.startsWith('/') || value.startsWith('@tarojs') || value.startsWith('react')) {
      return
    }
    scriptFiles.add(value)
  }
}

export const incrementId = () => {
  let n = 0
  return () => (n++).toString()
}

// 处理三方库引用
export function handleThirdPartyLib (filePath: string, nodePath: string[], root: string, convertRoot: string) {
  // 默认使用node_modules中的三方库
  if (typeof nodePath === 'undefined') {
    nodePath = [NODE_MODULES]
  }

  try {
    let isThirdPartyLibExist = false
    for (const modulePath of nodePath) {
      const parts = filePath.split('/')
      const npmModulePath = path.resolve(root, modulePath, parts[0])
      if (fs.existsSync(npmModulePath)) {
        isThirdPartyLibExist = true
        // 转换后的三方库放在node_modules中
        const moduleConvertPath = path.resolve(convertRoot, NODE_MODULES, parts[0])
        if (!fs.existsSync(moduleConvertPath)) {
          copyFolderToTaro(npmModulePath, moduleConvertPath)
          printLog(processTypeEnum.COPY, '三方库', normalizePath(npmModulePath.replace(path.join(root, '/'), '')))
        }
        break
      }
    }

    if (!isThirdPartyLibExist) {
      console.log(chalk.red(`在[${nodePath.toString()}]中没有找到依赖的三方库${filePath}，请安装依赖后运行`))
    }
  } catch (error) {
    console.log(chalk.red(`转换三方库${filePath}异常，请手动处理, error message:${error}`))
  }
}

export function getMatchUnconvertDir (importPath: string, externalPaths: string[]) {
  if (importPath === undefined || externalPaths === undefined) {
    return null
  }
  // 支持用户在convert.config.json中配置不转换的目录
  for (let externalPath of externalPaths) {
    externalPath = normalizePath(externalPath).replace('*', '.*')
    const reg = new RegExp(externalPath)
    const match = reg.exec(normalizePath(importPath))
    // external字段配置如 1：../demo 2：../demo/* 3：../demo/*/demo
    if (match) {
      // 处理../demo/*
      if (externalPath.length >= 2 && externalPath.endsWith('*')) {
        // 去掉结尾的".*"
        return externalPath.slice(0, externalPath.length - 3)
      }
      return match[0]
    }
  }
  return null
}

export function handleUnconvertDir (matchUnconvertDir: string, rootPath: string, convertRoot: string) {
  if (matchUnconvertDir == null) {
    return
  }
  // 支持用户在convert.config.json中配置不转换的目录
  const outputFilePath = matchUnconvertDir.replace(normalizePath(rootPath), normalizePath(convertRoot))
  if (!fs.existsSync(outputFilePath)) {
    copyFileToTaro(matchUnconvertDir, outputFilePath)
    printLog(processTypeEnum.COPY, '不转换目录', matchUnconvertDir.replace(normalizePath(path.join(rootPath, '/')), ''))
  }
}

export function getMatchDirPath (filePath: string, matchStr: string) {
  const reg = new RegExp(matchStr)
  const match = reg.exec(filePath)
  if (match) {
    if (matchStr.length >= 2 && matchStr.endsWith('*')) {
      return matchStr.slice(0, matchStr.length - 2)
    }
    return match[0]
  }
}

// 路径标准化
function normalizePath (path) {
  if (typeof path === 'undefined') {
    return ''
  }
  return path.replace(/\\/g, '/')
}

// 将数组中的相对路径转换为绝对路径
export function transRelToAbsPath (baseFilePath: string, fileRelativePaths: string[]) {
  if (typeof fileRelativePaths === 'undefined') {
    return []
  }
  const absolutePath: string[] = []
  for (const fileRelativePath of fileRelativePaths) {
    // 相对路径转为绝对路径
    absolutePath.push(path.resolve(baseFilePath, '..', fileRelativePath))
  }
  return absolutePath
}

// 获取wxss中引用文件的相对路径集合
export function getWxssImports (content: string) {
  if (content == null) {
    return []
  }
  // 匹配 /* ... */ 形式的注释
  const regex = /\/\*([\s\S]*?)\*\//g
  // 去掉css中的注释
  const contentWithoutComment = content.replace(regex, '')

  let match
  const imports: string[] = []
  const cssImportReg = new RegExp(CSS_IMPORT_REG)
  while ((match = cssImportReg.exec(contentWithoutComment)) !== null) {
    imports.push(match[2])
  }
  return imports
}

/**
 * copyFileTo： 将报告模版文件复制到转换后 taroConvert 目录中
 *
 * @param { string } sourceFilePath 源文件路径
 * @param { string } targeFileDir 转换后文件所在目录
 * @param { string } targeFileName 转换后文件名
 * @param { IReportMsg[] } reportErroMsg 报错信息
 */
export function generateReportFile (sourceFilePath, targeFileDir, targeFileName, reportErroMsg?: IReportMsg[]) {
  try {
    if (!fs.existsSync(targeFileDir)) {
      fs.mkdirSync(targeFileDir)
    }
    let data = fs.readFileSync(sourceFilePath, 'utf-8')
    if (reportErroMsg) {
      data = data.replace('__errorMsgReport__', JSON.stringify(reportErroMsg))
    }
    fs.writeFileSync(path.join(targeFileDir, targeFileName), data)
  } catch (error) {
    console.log(`文件${sourceFilePath}写入失败，errorMsg：${error}`)
  }
}

// eslint-disable-next-line camelcase
export const DEFAULT_Component_SET = new Set<string>([
  'View',
  'Icon',
  'Progress',
  'RichText',
  'Text',
  'Button',
  'Checkbox',
  'CheckboxGroup',
  'Form',
  'Input',
  'Label',
  'Picker',
  'PickerView',
  'PickerViewColumn',
  'Radio',
  'RadioGroup',
  'Slider',
  'Switch',
  'CoverImage',
  'Textarea',
  'CoverView',
  'MovableArea',
  'MovableView',
  'ScrollView',
  'Swiper',
  'SwiperItem',
  'Navigator',
  'Audio',
  'Camera',
  'Image',
  'LivePlayer',
  'Video',
  'Canvas',
  'Ad',
  'WebView',
  'Block',
  'Map',
  'Slot',
  'SlotView',
  'Editor',
  'MatchMedia',
  'FunctionalPageNavigator',
  'LivePusher',
  'OfficialAccount',
  'OpenData',
  'NavigationBar',
  'PageMeta',
  'VoipRoom',
  'AdCustom',
])
