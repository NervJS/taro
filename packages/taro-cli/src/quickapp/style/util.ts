export const comment = (msg, url) => 'Taro comment: ' + msg + '.' + (url ? '[兼容写法参考](' + url + ')' : '')
/**
 * 在rule中删除指定declaration
 */
export const removeDeclaration = (declaration, rule) => rule.declarations.splice(rule.declarations.findIndex(v => v.property === declaration.property && v.value === declaration.value), 1)
/**
 * 在rule中增加一条declaration
 */
export const addDeclaration = (property, value, rule) => rule.declarations.push({
  type: 'declaration',
  property,
  value
})
/**
 * 在rule中增加一条comment
 */
export const addComment = (property, value, rule) => rule.declarations.push({
  type: 'comment',
  comment: `${property}:${value}; ${comment('unsupported style', 'https://doc.quickapp.cn/widgets/common-styles.html')}`
})

export const getDeclarationValue = (property, rule) => {
  const declarations = rule.declarations.filter(declaration => declaration.property === property)
  return declarations.length && declarations[0] || false
}
