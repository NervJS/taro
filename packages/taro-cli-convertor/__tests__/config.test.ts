import Convertor from '../src'
import { clearMockFiles,normalizePath,resFileMap, setMockFiles  } from './__mocks__/fs-extra'
import { DEMO_JS_FILE_INFO, root } from './data/fileData'
import { removeBackslashesSerializer } from './util'

expect.addSnapshotSerializer(removeBackslashesSerializer)

const pathSystem = require('path')

jest.mock('fs', () => {
  const originalModule = jest.requireActual('fs') 
  return {
    ...originalModule,
    promises: {
      ...originalModule.promises,
      writeFile: jest.fn().mockImplementation((path,data)=>{
        let resPath = ''
        if(pathSystem.isAbsolute(path)){
          resPath = normalizePath(path).split(root)[1]
        }
        if(Buffer.isBuffer(data)){
          data = Buffer.from(data).toString('utf8')
        }
        resFileMap.set(root + resPath, data)
        console.log(resFileMap,'*************mock的fs操作*************')
      })
    },
  }
})

describe('小程序转换生成配置文件', () => {
  beforeAll(()=>{
    jest.spyOn(Convertor.prototype, 'init').mockImplementation(() => {})
  })

  afterEach(() => {
    // 清空文件信息
    clearMockFiles()
    jest.restoreAllMocks()
  })
  
  test('配置文件生成', (done) => {
    setMockFiles(root, DEMO_JS_FILE_INFO)
    const convertor = new Convertor(root, false)
    convertor.framework = 'react'
    convertor.generateConfigFiles()
    setTimeout(() => {
      console.log(resFileMap,'/////////////////打印test结果/////////////////')
      expect(resFileMap).toMatchSnapshot()
      done() 
    }, 4000) 
  })
})

