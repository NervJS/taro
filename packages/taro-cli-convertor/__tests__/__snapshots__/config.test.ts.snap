// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`小程序转换生成配置文件 配置文件生成 1`] = `
Map {
  /wxProject/taroConvert/package.json => {
  "name": "taroConvert",
  "version": "1.0.0",
  "private": true,
  "description": "",
  "templateInfo": {
    "name": "default",
    "typescript": false,
    "css": "sass"
  },
  "scripts": {
    "build:weapp": "taro build --type weapp",
    "build:swan": "taro build --type swan",
    "build:alipay": "taro build --type alipay",
    "build:tt": "taro build --type tt",
    "build:h5": "taro build --type h5",
    "build:rn": "taro build --type rn",
    "build:qq": "taro build --type qq",
    "build:jd": "taro build --type jd",
    "build:quickapp": "taro build --type quickapp",
    "build:mpharmony": "taro build --type mpharmony",
    "build:mpandroid": "taro build --type mpandroid",
    "dev:weapp": "npm run build:weapp -- --watch",
    "dev:swan": "npm run build:swan -- --watch",
    "dev:alipay": "npm run build:alipay -- --watch",
    "dev:tt": "npm run build:tt -- --watch",
    "dev:h5": "npm run build:h5 -- --watch",
    "dev:rn": "npm run build:rn -- --watch",
    "dev:qq": "npm run build:qq -- --watch",
    "dev:jd": "npm run build:jd -- --watch",
    "dev:quickapp": "npm run build:quickapp -- --watch",
    "dev:mpharmony": "npm run build:mpharmony -- --watch",
    "dev:mpandroid": "npm run build:mpandroid -- --watch",
    "test": "jest"
  },
  "browserslist": [
    "last 3 versions",
    "Android >= 4.1",
    "ios >= 8"
  ],
  "author": "",
  "dependencies": {
    "@babel/runtime": "^7.21.5",
    "@tarojs/components": "3.6.15",
    "@tarojs/helper": "3.6.15",
    "@tarojs/plugin-platform-weapp": "3.6.15",
    "@tarojs/plugin-platform-alipay": "3.6.15",
    "@tarojs/plugin-platform-tt": "3.6.15",
    "@tarojs/plugin-platform-swan": "3.6.15",
    "@tarojs/plugin-platform-jd": "3.6.15",
    "@tarojs/plugin-platform-qq": "3.6.15",
    "@tarojs/plugin-platform-h5": "3.6.15",
    "@tarojs/plugin-platform-mpharmony": "3.6.15",
    "@tarojs/plugin-platform-mpandroid": "3.6.15",
    "@tarojs/runtime": "3.6.15",
    "@tarojs/shared": "3.6.15",
    "@tarojs/taro": "3.6.15",
    "@tarojs/plugin-framework-react": "3.6.15",
    "@tarojs/react": "3.6.15",
    "react-dom": "^18.0.0",
    "react": "^18.0.0"
  },
  "devDependencies": {
    "@babel/core": "^7.8.0",
    "@tarojs/cli": "3.6.15",
    "@types/webpack-env": "^1.13.6",
    "@tarojs/test-utils-react": "^0.1.1",
    "@types/react": "^18.0.0",
    "webpack": "5.78.0",
    "@tarojs/taro-loader": "3.6.15",
    "@tarojs/webpack5-runner": "3.6.15",
    "babel-preset-taro": "3.6.15",
    "eslint-config-taro": "3.6.15",
    "eslint": "^8.12.0",
    "@pmmmwh/react-refresh-webpack-plugin": "^0.5.5",
    "react-refresh": "^0.11.0",
    "eslint-plugin-react": "^7.8.2",
    "eslint-plugin-import": "^2.12.0",
    "eslint-plugin-react-hooks": "^4.2.0",
    "stylelint": "^14.4.0",
    "postcss": "^8.4.18",
    "ts-node": "^10.9.1",
    "@types/node": "^18.15.11",
    "@types/jest": "^29.3.1",
    "jest": "^29.3.1",
    "jest-environment-jsdom": "^29.5.0"
  }
}
,
  /wxProject/taroConvert/config/index.js => import { defineConfig } from '@tarojs/cli'

import devConfig from './dev'
import prodConfig from './prod'

// https://taro-docs.jd.com/docs/next/config#defineconfig-辅助函数
export default defineConfig(async (merge, { command, mode }) => {
  const baseConfig = {
    projectName: 'taroConvert',
    date: '2023-12-28',
    designWidth: 750,
    deviceRatio: {
      640: 2.34 / 2,
      750: 1,
      375: 2,
      828: 1.81 / 2
    },
    sourceRoot: 'src',
    outputRoot: 'dist',
    plugins: [],
    defineConstants: {
    },
    copy: {
      patterns: [
      ],
      options: {
      }
    },
    framework: 'react',
    compiler: 'webpack5',
    cache: {
      enable: false // Webpack 持久化缓存配置，建议开启。默认配置请参考：https://docs.taro.zone/docs/config-detail#cache
    },
    mini: {
      postcss: {
        pxtransform: {
          enable: true,
          config: {

          }
        },
        url: {
          enable: true,
          config: {
            limit: 1024 // 设定转换尺寸上限
          }
        },
        cssModules: {
          enable: false, // 默认为 false，如需使用 css modules 功能，则设为 true
          config: {
            namingPattern: 'module', // 转换模式，取值为 global/module
            generateScopedName: '[name]__[local]___[hash:base64:5]'
          }
        }
      }
    },
    h5: {
      publicPath: '/',
      staticDirectory: 'static',
      output: {
        filename: 'js/[name].[hash:8].js',
        chunkFilename: 'js/[name].[chunkhash:8].js'
      },
      miniCssExtractPluginOption: {
        ignoreOrder: true,
        filename: 'css/[name].[hash].css',
        chunkFilename: 'css/[name].[chunkhash].css'
      },
      postcss: {
        autoprefixer: {
          enable: true,
          config: {}
        },
        cssModules: {
          enable: false, // 默认为 false，如需使用 css modules 功能，则设为 true
          config: {
            namingPattern: 'module', // 转换模式，取值为 global/module
            generateScopedName: '[name]__[local]___[hash:base64:5]'
          }
        }
      }
    },
    rn: {
      appName: 'taroDemo',
      postcss: {
        cssModules: {
          enable: false, // 默认为 false，如需使用 css modules 功能，则设为 true
        }
      }
    }
  }
  if (process.env.NODE_ENV === 'development') {
    // 本地开发构建配置（不混淆压缩）
    return merge({}, baseConfig, devConfig)
  }
  // 生产构建配置（默认开启压缩混淆等）
  return merge({}, baseConfig, prodConfig)
})
,
}
`;
